components/f4jumble/src/lib.rs

```diff
@@ -1,13 +1,28 @@
+#![no_std]
+
 use blake2b_simd::{Params as Blake2bParams, OUTBYTES};
-use std::cmp::min;
-use std::ops::RangeInclusive;
+use const_format::formatcp;
+use core::cmp::min;
+use core::ops::RangeInclusive;
+use core::result::Result;
+
+#[cfg(feature = "std")]
+extern crate std;
+#[cfg(feature = "std")]
+use std::vec::Vec;
 
 #[cfg(test)]
 mod test_vectors;
 #[cfg(test)]
 mod test_vectors_long;
 
-const VALID_LENGTH: RangeInclusive<usize> = 48..=4194368;
+pub const VALID_LENGTH: RangeInclusive<usize> = 48..=4194368;
+const INVALID_LENGTH_ERROR_MESSAGE: &str = formatcp!(
+    "Message length must be in interval ({}..={})",
+    *VALID_LENGTH.start(),
+    *VALID_LENGTH.end()
+);
+//more andvanced formatcp! only in nightly Rust
```

Comment:
Rather than using a string (which is hard to keep track of in callers), we should use an `Error` enum with a single case. Then if `std` is enabled, we can implement `std::error::Error` for it.

We should also return this `Error` enum from the allocating methods.

Reply:
Something like:
```suggestion
use core::fmt;

#[derive(Debug)]
pub enum Error {
    InvalidLength,
}

impl fmt::Display for Error {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::InvalidLength => [...],
        }
    }
}

#[cfg(feature = "std")]
impl std::error::Error for Error {}
```